packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts

```diff
@@ -178,12 +119,95 @@ function rebaseMarkList<TNodeChange>(
+
+    public pop(): RebaseMarks<T> {
+        const baseMark = this.baseMarks.peek();
+        const newMark = this.newMarks.peek();
+
+        if (baseMark === undefined || newMark === undefined) {
+            return {
+                baseMark: this.baseMarks.pop(),
+                newMark: this.newMarks.pop(),
+                inverseRevision: this.reattachRevision,
+            };
+        } else if (isAttach(baseMark) && isAttach(newMark)) {
+            const revision = baseMark.revision ?? this.baseRevision;
+            const reattachOffset = getOffsetInReattach(newMark.lineage, revision);
+            if (reattachOffset !== undefined) {
+                const offset = reattachOffset - this.reattachOffset;
+                this.reattachOffset = reattachOffset;
+                this.reattachRevision = revision;
+                if (offset === 0) {
+                    return { newMark: this.newMarks.pop(), inverseRevision: this.reattachRevision };
+                } else if (offset >= getOutputLength(baseMark)) {
+                    return { baseMark: this.baseMarks.pop() };
+                } else {
+                    const [baseMark1, baseMark2] = splitMarkOnOutput(baseMark, offset);
+                    this.baseMarks.push(baseMark2);
+                    return { baseMark: baseMark1 };
+                }
+            } else if (isAttachAfterBaseAttach(newMark, baseMark)) {
+                return { baseMark: this.baseMarks.pop() };
+            } else {
+                return { newMark: this.newMarks.pop(), inverseRevision: this.reattachRevision };
+            }
+        } else if (isAttach(newMark)) {
+            return { newMark: this.newMarks.pop(), inverseRevision: this.reattachRevision };
+        }
+
+        // TODO: Handle case where base has adjacent or nested inverse reattaches from multiple revisions
+        this.reattachOffset = 0;
+        this.reattachRevision = undefined;
+        if (isAttach(baseMark)) {
+            return { baseMark: this.baseMarks.pop() };
+        } else {
+            this.reattachOffset = 0;
+            this.baseMarks.pop();
+            this.newMarks.pop();
+            const newMarkLength = getInputLength(newMark);
+            const baseMarkLength = getInputLength(baseMark);
+            if (newMarkLength < baseMarkLength) {
+                const [baseMark1, baseMark2] = splitMarkOnInput(baseMark, newMarkLength);
+                this.baseMarks.push(baseMark2);
+                return { baseMark: baseMark1, newMark };
+            } else if (newMarkLength > baseMarkLength) {
+                const [newMark1, newMark2] = splitMarkOnInput(newMark, baseMarkLength);
+                this.newMarks.push(newMark2);
+                return { baseMark, newMark: newMark1 };
+            } else {
+                return { baseMark, newMark };
+            }
+        }
+    }
+}
+
+interface RebaseMarks<T> {
+    baseMark?: Mark<T>;
+    newMark?: Mark<T>;
+    inverseRevision?: RevisionTag;
```

Comment:
Would be nice to document when this is populated and what it contains when it is.